satellite/gc/service.go

```diff
@@ -0,0 +1,198 @@
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int64   `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+	ConcurrentSends   int64   `help:"the number of nodes to concurrently send bloom filters to" releaseDefault:"1" devDefault:"1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log              *zap.Logger
+	Loop             *sync2.Cycle
+	metainfoloop     *metainfo.Loop
+	transport        transport.Client
+	overlay          overlay.DB
+	lastSendTime     time.Time
+	lastPieceCounts  atomic.Value
+	config           Config
+	errGroupMutex    *sync.Mutex
+	pieceCountsMutex *sync.Mutex
+}
+
+// RetainInfo contains info needed for a storage node to retain important data and delete garbage data
+type RetainInfo struct {
+	Filter       *bloomfilter.Filter
+	CreationDate time.Time
+	count        int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, transport transport.Client, overlay overlay.DB, loop *metainfo.Loop, config Config) *Service {
+	// TODO retrieve piece counts from overlay (when there is a column for them)
+	var lastPieceCounts atomic.Value
+	lastPieceCounts.Store(map[storj.NodeID]int{})
+
+	return &Service{
+		log:              log,
+		Loop:             sync2.NewCycle(config.Interval),
+		metainfoloop:     loop,
+		transport:        transport,
+		overlay:          overlay,
+		lastPieceCounts:  lastPieceCounts,
+		config:           config,
+		errGroupMutex:    &sync.Mutex{},
+		pieceCountsMutex: &sync.Mutex{},
+	}
+}
+
+// Run starts the gc loop service
+func (service *Service) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	return service.Loop.Run(ctx, func(ctx context.Context) error {
+		pieceCounts := service.lastPieceCountsValue()
+		obs := NewObserver(service.log.Named("gc observer"), pieceCounts, service.config)
+
+		err := service.metainfoloop.Join(ctx, obs)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = service.Send(ctx, obs)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		return nil
+	})
+}
+
+// Send sends the piece retain requests to all storage nodes
+func (service *Service) Send(ctx context.Context, obs *Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	service.lastSendTime = time.Now().UTC()
+	newPieceCounts := make(map[storj.NodeID]int)
+
+	limiter := sync2.NewLimiter(int(service.config.ConcurrentSends))
+	var errList errs.Group
+	for id, retainInfo := range obs.retainInfos {
+		service.sendRetainFromLimiter(ctx, id, retainInfo, newPieceCounts, limiter, errList)
+	}
+	limiter.Wait()
+	if errList.Err() != nil {
+		service.log.Error("error sending retain infos", zap.Error(errList.Err()))
```

Comment:
Part of me says "if we're just going to log these, why did we do all this fancy errGroup with mutex junk"

Reply:
True